##  SENTIMENT ANALYSIS FOR KENYA AIRWAYS

###  AUTHORS

* Winfred Kinya
* Jeremiah Waiguru
* Paul Ngatia
* Mercy Kiragu

##  1.0  PROJECT OVERVIEW

This project is focused on analyzing customer reviews of Kenya Airways to extract meaningful insights regarding customer satisfaction and service quality. By utilizing sentiment analysis techniques, the project will classify these reviews into positive, negative, or neutral sentiments. The goal is to help Kenya Airways identify areas of excellence as well as aspects of their service that may require improvement. This analysis will serve as a foundation for enhancing customer experience and driving operational improvements.

##  1.1  BUSINESS UNDERSTANDING

In the competitive aviation industry, customer satisfaction plays a vital role in shaping the reputation and success of an airline. Kenya Airways, like many airlines, receives numerous reviews from passengers, which contain valuable feedback about their experiences. Understanding the sentiments behind these reviews is essential for identifying strengths and weaknesses in service delivery. This project will leverage sentiment analysis to process and analyze customer feedback, enabling Kenya Airways to make informed decisions that enhance passenger satisfaction and loyalty.

##  1.2  PROBLEM STATEMENT

Kenya Airways is challenged by the vast amount of unstructured customer feedback available through online reviews. Without an efficient method to analyze this feedback, the airline may overlook critical insights that could lead to service improvements. This project aims to address this challenge by systematically analyzing the sentiment of customer reviews, providing Kenya Airways with a clear understanding of customer perceptions and identifying key areas that require attention.

##  1.3  OBJECTIVES

### Primary objective

To develop a robust sentiment analysis model that accurately classifies customer reviews of Kenya Airways into positive, negative, or neutral sentiments.

###  Specific objectives

1. To explore the sentiment distribution across different customer demographics, travel classes, and time periods.
2. To identify recurring themes and specific aspects of service (e.g., customer service, punctuality, in-flight experience) that are frequently mentioned in the reviews.
3. To assess the performance of various sentiment analysis models and determine the most effective model for this dataset.
4. To generate actionable insights from the sentiment analysis that can be used by Kenya Airways to improve customer service and address common issues

##  2.0  DATA UNDERSTANDING

We did Web Scraping from the 3 websites below and merged our datasets

1. https://uk.trustpilot.com/review/www.kenya-airways.com
2. https://www.airlinequality.com/airline-reviews/kenya-airways/
3. https://www.airlineratings.com/airlines/kenya-airways

The dataset has 3065 rows and 5 columns with the following description ;

* Departures: This column indicates the departure region or type of flight, such as "Africa" or "International."
* Class: This column specifies the class of travel, such as "Economy" or "Business Class."
* Review: This column contains customer reviews of their experiences with Kenya Airways. The text data may include opinions, complaints, and compliments about the airline.
* Rating: This column contains the numeric rating provided by the customer, typically ranging from 1.0 to 5.0, where 1.0 represents "Poor" and 5.0 represents "Excellent."
* Rating_description: This column provides a textual description corresponding to the numeric rating, such as "Poor," "Good," or "Excellent."

## Importing Necessary Libraries

In [58]:
# importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use("fivethirtyeight")
import seaborn as sns
import plotly.express as px 
import re
import string
import joblib

# ignore warnings
import warnings
warnings.filterwarnings("ignore")

from sklearn.metrics import precision_score, accuracy_score, recall_score, f1_score, classification_report
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_extraction.text import TfidfVectorizer
from wordcloud import WordCloud 
from sklearn.metrics import roc_curve, auc

#downloading dependencies
import nltk
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
nltk.download('vader_lexicon')

from nltk.corpus import stopwords
from nltk.sentiment.vader import SentimentIntensityAnalyzer
from nltk.stem import SnowballStemmer
from nltk.stem import WordNetLemmatizer
from nltk import pos_tag
from nltk.corpus import wordnet


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\PC\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\PC\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\PC\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\PC\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\PC\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


## 2. 1 Loading Datasets

### 1.Dataset one

In [59]:
#Kenya_airways_flights
df = pd.read_csv('Refined data/Dataset_one.csv')
df.head()

Date3     Departures           Class  \
0  2022-09-01         Africa         Economy   
1  2022-11-01  International         Economy   
2  2022-05-01  International         Economy   
3  2022-11-01         Africa  Business Class   
4  2022-11-01         Africa         Economy   

                                              Review  Rating_Value  \
0  Recently I travelled for business from Cape To...           4.0   
1  I want to thanks to Jane for her excellent ser...           5.0   
2  not a bad airline to fly with could do with a ...           4.0   
3  Dear Kenya Airways ,\r\n\r\nI am on my way to ...           1.0   
4  Absolutely appalling airline, never ever use i...           1.0   

  Rating_Description  
0               Good  
1          Excellent  
2               Good  
3               Poor  
4               Poor

In [60]:
def describe_columns(df):
    # Print column names
    print("Column Names:")
    print(df.columns)

    # Print number of rows and columns
    print("\nShape:")
    print(df.shape)

describe_columns(df)

Column Names:
Index(['Date3', 'Departures', 'Class', 'Review', 'Rating_Value',
       'Rating_Description'],
      dtype='object')

Shape:
(2345, 6)


The Dataset has 2345 rows and 6 columns.
The other columns are categorical or text-based.

### 2.Dataset two

In [61]:
#Kenya_airways_reviews
df = pd.read_csv('Refined data/Dataset_two.csv')
df.head()

Date3                                             Review
0  2024-02-24  ✅ Trip Verified | I have never seen such a dis...
1  2024-02-24  Not Verified |   Bumped off flight at boarding...
2  2024-07-30  ✅ Trip Verified |   I am disappointment with t...
3  2024-07-25  ✅ Trip Verified |  This airline is becoming mo...
4  2024-07-24  ✅ Trip Verified |  3 out of 4 flights delayed ...

In [62]:
def describe_columns(df):
    # Print column names
    print("Column Names:")
    print(df.columns)

    # Print number of rows and columns 
    print("\nShape:")
    print(df.shape)

describe_columns(df)

Column Names:
Index(['Date3', 'Review'], dtype='object')

Shape:
(477, 2)


The Dataset has 477 rows and 2 columns.

### 3.Dataset three

In [63]:
# KQ_reviews
df = pd.read_csv('Refined data/Dataset_three.csv')
df.head()

Date3  Rating                                             Review
0  02-08-2024     0.0  Constantly canceling flights Kenya airlines ca...
1  04-08-2024     0.0  Kenya Airways canceled the flight and… Kenya A...
2  30-07-2024     0.0  If your thinking of Kenya Airlines...think aga...
3  29-07-2024     0.0  Worst airlines ever Worst airlines ever, not f...
4  25-07-2024     0.0  We were trying to go to Nampula from Nairobi a...

In [64]:
def describe_columns(df):
    # Print column names
    print("Column Names:")
    print(df.columns)

    # Print number of rows and columns 
    print("\nShape:")
    print(df.shape)

describe_columns(df)

Column Names:
Index(['Date3', 'Rating', 'Review'], dtype='object')

Shape:
(243, 3)


The Dataset has 243 rows and 3 columns.Mainly, 'Date', 'Rating and 'Review'

## Loading the Merged Dataset

In [65]:
# KQ_Data
df = pd.read_csv('Refined data/KQ_Data.csv')
df

Date3     Departures           Class  \
0     2022-09-01 00:00:00         Africa         Economy   
1     2022-11-01 00:00:00  International         Economy   
2     2022-05-01 00:00:00  International         Economy   
3     2022-11-01 00:00:00         Africa  Business Class   
4     2022-11-01 00:00:00         Africa         Economy   
...                   ...            ...             ...   
3060           26-01-2017            NaN             NaN   
3061           30-08-2016            NaN             NaN   
3062           02-07-2016            NaN             NaN   
3063           21-11-2015            NaN             NaN   
3064           02-07-2014            NaN             NaN   

                                                 Review  Rating  \
0     Recently I travelled for business from Cape To...     4.0   
1     I want to thanks to Jane for her excellent ser...     5.0   
2     not a bad airline to fly with could do with a ...     4.0   
3     Dear Kenya Airways ,\r\n\r\nI am on my way to ...     1.0   
4     Absolutely appalling airline, never ever use i...     1.0   
...                                                 ...     ...   
3060  Better than I expected Much better service tha...     0.0   
3061  Request for immediate action. I have never had...     0.0   
3062  Never again Flew with Kenya airways last year ...     0.0   
3063  NEVER AGAIN - RUBISH COMPANY !!! We travelled ...     1.0   
3064  Worse Customer service ever!!! Will never ever...     0.0   

     Rating_Description  
0                  Good  
1             Excellent  
2                  Good  
3                  Poor  
4                  Poor  
...                 ...  
3060                NaN  
3061                NaN  
3062                NaN  
3063                NaN  
3064                NaN  

[3065 rows x 6 columns]

In [66]:
def describe_columns(df):
    # Print column names
    print("Column Names:")
    print(df.columns)

    # Print data types
    print("\nData Types:")
    print(df.dtypes)

    # Print number of rows and columns
    print("\nShape:")
    print(df.shape)

    # Print df information
    print("\nInfo:")
    print(df.info())

    # Print descriptive statistics for numerical columns
    print("\nDescriptive Statistics:")
    print(df.describe())

    # Print missing values count per column
    print("\nMissing Values in percentages:")
    print((df.isna().sum()/len(df)) * 100)

describe_columns(df)

Column Names:
Index(['Date3', 'Departures', 'Class', 'Review', 'Rating',
       'Rating_Description'],
      dtype='object')

Data Types:
Date3                  object
Departures             object
Class                  object
Review                 object
Rating                float64
Rating_Description     object
dtype: object

Shape:
(3065, 6)

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3065 entries, 0 to 3064
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Date3               3063 non-null   object 
 1   Departures          2345 non-null   object 
 2   Class               2345 non-null   object 
 3   Review              3065 non-null   object 
 4   Rating              2578 non-null   float64
 5   Rating_Description  2345 non-null   object 
dtypes: float64(1), object(5)
memory usage: 143.8+ KB
None

Descriptive Statistics:
            Rating
count  2578.000000
mean      2.865012
std 

The dataset has 3065 rows and 6 columns. It also has missing values, with some columns registering aproximately 90%. This will be dealt with at the preprocessing stage. From our data overview, the Review column needs preprocessing since it contains raw text with numbers,nametags, emojis and punctuations.